First draft of problems document (#63)
capi-workgroup/problems: capi_problems.rst

Over the decades, the C API evolved to provide different tiers of stability,
conventions changed, and new usage patterns have emerged, such as bindings
to languages other than C/C++. In addition, lessons were learned and
mistakes in both the design and the implementation of the C API were
identified.

Evolving the C API is hard due to the combination of backwards
compatibility constraints and its inherent complexity, both
technical and social. Different types of users bring different,
sometimes conflicting, requirements. The tradeoff between stability
and progress is an ongoing, highly contentious topic of discussion
when suggestions are made for incremental improvements.
Several proposals have been put forward for improvement, redesign
or replacement of the C API, each representing a deep analysis of
the problems.  At the 2023 Language Summit, three back-to-back
sessions were devoted to different aspects of the C API. There is
general agreement that a new design can remedy the problems that
the C API has accumulated over the last 30 years, while at the same
time updating it for use cases that it was not originally designed for.

> However, there was a also sense at the Language Summit that we are

Reviewer: ```suggestion
However, there was also a sense at the Language Summit that we are
```

and the particular requirements that each of them has.

C API Stakeholders
==================

[TODO: write this section]

API Evolution and Maintenance
=============================

The difficulty of making changes in the C API is central to this report. It is
implicit in many of the issues we discuss here, particularly when we need to
decide whether an incremental bugfix can resolve the issue, or whether it can
only be addressed as part of an API redesign
[`Issue 44 <https://github.com/capi-workgroup/problems/issues/44>`__]. The
benefit of each incremental change is often viewed as too small to justify the
disruption. Over time, this implies that every mistake we make in an API's
design or implementation remains with us indefinitely.

We can take two views on this issue. One is that this is a problem and the
solution needs to be baked into any new C API we design, in the form of a
process for incremental API evolution. The other possible approach is that
this is not a problem to be solved, but rather a feature of any API. In this
view, API evolution should not be incremental, but rather through large
redesigns, each of which learns from the mistakes of the past and is not

> shackled by backwards compatibility requirements. A realistic approach

Reviewer: This reads a bit too normative to me, inconsistent with the document's stated goal of listing only problems, not solutions.
```suggestion
shackled by backwards compatibility requirements. A compromise approach
```

or important enough to tackle incrementally, and leaving others alone.

The problem we have in CPython is that we don't have an agreed, official
approach to API evolution. Different members of the core team are pulling in
different directions and this is an ongoing source of disagreements.
Any new C API needs to come with a clear decision about the model
that its maintenance will follow, as well as the technical and
organizational processes by which this will work.

If the model does include provisions for incremental evolution of the API,
it will include processes for managing the impact of the change on users
[`Issue 60 <https://github.com/capi-workgroup/problems/issues/60>`__],
perhaps through introducing an external backwards compatibility module
[`Issue 62 <https://github.com/capi-workgroup/problems/issues/62>`__],
or a new API tier of "blessed" functions
[`Issue 55 <https://github.com/capi-workgroup/problems/issues/55>`__].

API Specification and Abstraction
=================================

The C API does not have a formal specification, it is described
semi-formally in the documentation and exposed through C header
files. This creates a number of problems.

> Bindings for languages other than C/C++ are required to parse C code

Reviewer: ```suggestion
Bindings for languages other than C/C++ must parse C code
```
Current wording sounds like in order to parse C code, you need bindings for other languages, when in fact it's the reverse.

problem.

There are functions that take a ``PyObject*`` argument, with special meaning
when it is ``NULL``. For example, if ``PyObject_SetAttr`` receives ``NULL`` as
the value to set, this mean that the attribute should be cleared. This is error
prone because it could be that ``NULL`` indicates an error in the construction
of the value, and the program failed to check for this error. The program will
misinterpret the ``NULL`` to mean something different than error
[`Issue 47 <https://github.com/capi-workgroup/problems/issues/47>`__].

API Tiers and Stability Guarantees
==================================

The different API tiers provide different tradeoffs of stability vs
API evolution, and sometimes performance.

The stable ABI was identified as an area that needs to be looked into. At
the moment it is incomplete and not widely adopted. At the same time, its
existence is making it hard to make changes to some implementation
details, because it exposes struct fields such as ``ob_refcnt``,
``ob_type`` and ``ob_size``. There was some discussion about whether
the stable ABI is worth keeping. Arguments on both sides can be

> found in `Issue 4 <https://github.com/capi-workgroup/problems/issues/4>`__
> and `Issue 9 <https://github.com/capi-workgroup/problems/issues/9>`__].

Reviewer 2: ```suggestion
found in [`Issue 4 <https://github.com/capi-workgroup/problems/issues/4>`__]
and [`Issue 9 <https://github.com/capi-workgroup/problems/issues/9>`__].
```

The limited API was introduced in 3.2 as a blessed subset of the C API
which is recommended for users who would like to restrict themselves
to high quality APIs which are not likely to change often. The
``Py_LIMITED_API`` flag allows users to restrict their program to older
versions of the limited API, but we now need the opposite option, to
exclude older versions. This would make it possible to evolve the
limited API by replacing flawed elements in it
[`Issue 54 <https://github.com/capi-workgroup/problems/issues/54>`__].
More generally, in a redesign we should revisit the way that API
tiers are specified and consider designing a method that will unify the
way we currently select between the different tiers
[`Issue 59 <https://github.com/capi-workgroup/problems/issues/59>`__].

API elements whose names begin with an underscore are considered
private, essentially an API tier with no stability guarantees.
However, this was only clarified recently, in
`PEP 689 <https://peps.python.org/pep-0689/>`__. It is not clear
what the change policy should be with respect to such API elements
that predate PEP 689
[`Issue 58 <https://github.com/capi-workgroup/problems/issues/58>`__].

> There are API functions which have a fast version and a safe version
> (like ``PyTuple_GET_ITEM`` vs ``PyTuple_GetItem``). It may help to
> be able to group them into their own tiers - the "fast API" tier and
> the "safe API" tier
> [`Issue 61 <https://github.com/capi-workgroup/problems/issues/61>`__].

Reviewer 2: I would prefer to avoid the "fast" term here, it can be misleading.

```suggestion
There are API functions which have an unsafe version (no error checking) and a safe version
(like ``PyTuple_GET_ITEM`` vs ``PyTuple_GetItem``). It may help to
be able to group them into their own tiers - the "unsafe API" tier and
the "safe API" tier
[`Issue 61 <https://github.com/capi-workgroup/problems/issues/61>`__].
```

The C Language
==============

A number of issues were raised with respect to the way that CPython
uses the C language. First there is the issue of which C dialect
we use, and how we test our compatibility with it
[`Issue 42 <https://github.com/capi-workgroup/problems/issues/42>`__].

Usage of ``const`` in the API is currently sparse, but it is not
clear whether this is something that we should consider changing
[`Issue 38 <https://github.com/capi-workgroup/problems/issues/38>`__].

We currently use the C types ``long`` and ``int``, where ``stdint``
and ``int32_t`` would have been better choices
[`Issue 27 <https://github.com/capi-workgroup/problems/issues/27>`__].

We are using C language features which are hard for other languages
to interact with
[`Issue 35 <https://github.com/capi-workgroup/problems/issues/35>`__].

There are API functions that take a ``PyObject*`` arg which must be
of a more specific type (such as ``PyTuple_Size``, which fails if

> its arg is not a ``PyTuple``). It is an open question whether this

Reviewer: ```suggestion
its arg is not a ``PyTupleObject*``). It is an open question whether this
```
Nit, but "PyTuple" isn't a type in either Python or C.

equivalents.  The behaviour of ``PyIter_Next`` is different from
``tp_iternext``.
[`Issue 29 <https://github.com/capi-workgroup/problems/issues/29>`__].
The behaviour of ``PySet_Contains`` is different from ``set.__contains__``
[`Issue 6 <https://github.com/capi-workgroup/problems/issues/6>`__].

The fact that ``PyArg_ParseTupleAndKeywords`` takes a non-const
char* array as argument makes it more difficult to use.
[`Issue 28 <https://github.com/capi-workgroup/problems/issues/28>`__].

Python.h does not expose the whole API. Some headers (like marshal.h)
are not included from Python.h.
[`Issue 43 <https://github.com/capi-workgroup/problems/issues/43>`__].

Missing Functionality
=====================

This section consists of a list of feature requests, i.e., functionality
that was identified as missing in the current C API.

**Debug Mode**

A debug mode that can be activated without recompilation and which
activates various checks that can help detect various types of errors.
[`Issue 36 <https://github.com/capi-workgroup/problems/issues/36>`__].

**Introspection**

There aren't currently reliable introspection capabilities for objects

> defined in C in the same way as there are for python objects.

Reviewer: ```suggestion
defined in C in the same way as there are for Python objects.
```
